eeps/eep-0064.md

```diff
@@ -0,0 +1,187 @@
+using string interpolation).
+
+### Binary-String Design
+
+To make the usage of binary-string similar to other languages, this EEP proposal
+makes the binary-string design decisions similar to the ones for the Elixir language.
+
+#### Binary-String Start and End
+
+Binary-strings must start and end with triple-quotes in their own lines:
+
+*Code Example*
+
+    X = """
+        This is the beginning of the triple-quote text
+        """
+
+*Documentation Example*
+
+    -doc """
+         Removes double-quotes (") from a given string
+         """
+    remove_double_quotes(X) -> 
+
+#### Binary-Strings Closing of Triple-Quotes
+
+The closing of the triple-quotes denotes the space (indentation):
+
+*1. Code Example*
+
+    >> X = """
+           This text
+           has no indentation
+           """
+    
+    Equivalent to: <<"This text\nhas no indentation\n">>
+
+*2. Code Example*
+
+    >> Y = """
+           This text
+           has indentation
+    """
+    
+    Equivalent to: <<"       This text\n       has indentation\n">>
+
+*3. Code Example*
+
+    >> Z = """
+      This text has
+      two space indentation
+    """
+    
+    Equivalent to: <<"  This text has\n  two space indentation\n">>
+
+*1. Documentation Example - No indentation*
+
+    -doc """
+         Removes double-quotes (") from a given string
+         """
+    remove_double_quotes(X) -> 
+
+*2. Documentation Example - No indentation*
+
+    -doc """
+    Removes double-quotes (") from a given string
+    """
+    remove_double_quotes(X) -> 
+
+*3. Documentation Example - Indentation*
+
+    -doc """
+         Removes double-quotes (") from a given string
+    """
+    remove_double_quotes(X) -> 
+
+#### Binary-Strings Errors
+
+An error should be raised when there is text that starts before the closing of triple-quotes, where
+the error makes the code written using triple-quotes to have a uniform style:
+
+*Code Example*
+
+    >> Err = """
+    This shall be an error
+             """
+   
+    error             
+
+*Documentation Example*
+
+    -doc """
+    Removes double-quotes (") from a given string
+         """
+    remove_double_quotes(X) -> 
+
+#### Escaping Newlines in Triple-Quotes
+
+Escaping new lines using `\`:
+
+To write `\` inside triple quotes, one needs to use `\\`. [EEP-59][] should
+ignore the meaning of `\` if inside a code block.
+
+*Code Example*
+
+    >> X = """
+           This text \
+           has no indentation
+           """
+    
+    Equivalent to: <<"This text has no indentation\n">>
+
+*Documentation Example*
+
+    -doc """
+    Removes double-quotes (") \
+    from a given string
+    """
+    remove_double_quotes(X) -> 
+
+[EEP 59]: https://www.erlang.org/eeps/eep-0059
+    "EEP 59: Module attributes for documentation"
+
+[EEP-62]: https://www.erlang.org/eeps/eep-0062
+    "String Interpolation Syntax"
+
+[PR-7343]: https://github.com/erlang/otp/pull/7343
+    "Feature: String Interpolation"
+
+Copyright
+=========
```

Comment:
Should there be a discussion about compatibility? In particular, `"""foo"""` is valid Erlang today. It is very unlikely, but perhaps worth mentioning.

Reply:
I think this discussion has diverged too far from "what problem do we want to solve" with triple-quoted binary strings.
For me the problem to solve first is "How to write the documentation (Markdown assumed) in the `-doc` attribute in a convenient way. This is a different context from the program code in the same module. The text written in the `-doc` attribute is not to be available for the "program" , it should just land in a "doc-chunk" .

Another separate problem we want to solve is to have a nicer syntax for utf8 encoded binaries from <<"abc">> to something like `b"abc"`.

There are other things we want to improve as well and there might be an advantage to combine them, reuse notation etc. But first I think we should solve how to write documentation. What is written inside the `-doc` attribute is not part of the program in my view it is like a comment with the special property "part of the documentation" and its content does not matter for the ordinary compiler.

Reply:
Maybe I'm missing something, but I don't see triple quoted strings as a necessary prerequisite for [EEP-59](https://www.erlang.org/eeps/eep-0059). Wouldn't the following suffice?

````erlang
-module(z).

-doc "
This is an example module doc.

With embedded preformatted text:

    f() ->
        ok.

And also backtick-delimited text, too:

```erlang
g() ->
    ok.
```
".

f() -> ok.
````

All the formatting is preserved, so it's up to the Markdown engine to interpret it:

````
20> z:module_info(attributes).
[{vsn,[105165884523284496141303209755187911974]},
 {doc,"\nThis is an example module doc.\n\nWith embedded preformatted text:\n\n    f() ->\n        ok.\n\nAnd also backtick-delimited text, too:\n\n```erlang\ng() ->\n    ok.\n```\n"}]
21> io:format("~ts\n", [string:trim(ModuleDoc)]).
This is an example module doc.

With embedded preformatted text:

    f() ->
        ok.

And also backtick-delimited text, too:

```erlang
g() ->
    ok.
```
ok
22>
````

Indeed, stripping leading whitespace could allow for more flexibility when formatting the documentation in source code (so triple-quoted strings could come in handy if/when they arrive), but it doesn't seem to be a must have for EEP-59.

Reply:
Here is the issue in my opinion. Imagine writing docs for `string:uppercase/1`:

```erlang
-doc "
Converts `String` to uppercase.

See also `titlecase/1`.

## Examples

    1> string:uppercase(\"Michał\").
    \"MICHAŁ\".

".
```

Basically all of the `string` module would be quite annoying to write examples for. But I agree with @KennethL. The issue is documentation. Triple-quoted strings is one of the possible solutions.